# library
library(dplyr)
library(taxizedb)
library(taxonomizr)
library(seqinr)
library(rentrez)
library(XML)
library(xlsx)
library(ggplot2)
library(stringr)

####
# Combine CCtyper and Makarova Databases ####
####

# get Makarova database entries
cctyperdb <- read.table("data/Makarova/accession_list.txt", header = FALSE, sep = " ")
colnames(cctyperdb) <- c("Accession","Hmm")
cctyperdb$Hmm <- gsub(".hmm","",cctyperdb$Hmm)
makarovadb <- read.table("data/Makarova/profFam.tab", header = FALSE, sep = "\t")
colnames(makarovadb) <- c("Accession","CasClass", "CasClassSubtype", "Desc")
cctyperdb <- cctyperdb %>% left_join(makarovadb) %>% select(-Desc)

# manage column formats and generate new columns
cctyperdb$CasClassSubtype <- gsub(",", "|", cctyperdb$CasClassSubtype)
cctyperdb$HmmClass <- sapply(cctyperdb$Hmm, function(x){
  strsplit(x, split = "_")[[1]][1]
})
cctyperdb$HmmClass2 <- cctyperdb$HmmClass
cctyperdb$HmmClass2[grepl("Cas|Csc|Csb|Csf|Cse|Csx|Csm|Cmr", cctyperdb$HmmClass)] <- str_extract(string = cctyperdb$HmmClass2[grepl("Cas|Csc|Csb|Csf|Cse|Csx|Csm|Cmr", cctyperdb$HmmClass2)], 
                                                                                    pattern = "(Cas|Csc|Csb|Csf|Cse|Csx|Csm|Cmr)[0-9]+")
cctyperdb$HmmClass2[cctyperdb$HmmClass=="CasR"] <- "CasR"
cctyperdb$HmmClass <- cctyperdb$HmmClass2
cctyperdb$HmmClass2 <- NULL
  
####
# Import and Combine RNA and DNA CCTyper Results ####
####

# combine DNA and RNA results from CCTyper
# These tab files are generated by the CCTyper Pipeline
trinity_DNA_hmmer <- read.table("data/Trinity_DNA_sense_hmmer.tab", header = TRUE, sep = "\t")
trinity_DNA_hmmer$ORF <- gsub("TRINITY", "TRINITY_DNA", trinity_DNA_hmmer$ORF)
trinity_DNA_hmmer$DataType <- "DNA"
trinity_RNA_hmmer <- read.table("data/Trinity_RNA_sense_hmmer.tab", header = TRUE, sep = "\t")
trinity_RNA_hmmer$ORF <- gsub("TRINITY", "TRINITY_RNA", trinity_RNA_hmmer$ORF)
trinity_RNA_hmmer$DataType <- "RNA"
trinity_hmmer <- rbind(trinity_RNA_hmmer, trinity_DNA_hmmer)

# combine hmmer results with database
trinity_hmmer <- trinity_hmmer %>% left_join(cctyperdb) %>% select(-Pos)
trinity_hmmer$Sample <- sapply(trinity_hmmer$ORF, function(x){
  temp <- strsplit(x, split = "_")[[1]]
  return(paste(temp[-length(temp)], collapse = "_"))
})

####
# Get RNA and DNA sample ORF sequences ####
####

RNA_fasta_selected <- readRDS("data/Trinity_RNA_sense_proteins_selected.rds")
RNA_fasta_selected$ORF <- gsub("TRINITY", "TRINITY_RNA", RNA_fasta_selected$ORF)
DNA_fasta_selected <- readRDS("data/Trinity_DNA_sense_proteins_selected.rds")
DNA_fasta_selected$ORF <- gsub("TRINITY", "TRINITY_DNA", DNA_fasta_selected$ORF)
fasta_selected <- rbind(RNA_fasta_selected, DNA_fasta_selected)
trinity_hmmer <- trinity_hmmer %>% left_join(fasta_selected, by = c("ORF" = "ORF"))

####
# Attach Sequences of Fasta Sequences ####
####

# get fasta files 
list_of_fasta <- list.files("data/Makarova/Supplementary_Dataset_2.profiles/", full.names = TRUE)
list_of_fasta <- list_of_fasta[grepl(".FASTA", list_of_fasta)]
list_of_fasta_names <- list.files("data/Makarova/Supplementary_Dataset_2.profiles/", full.names = FALSE)
list_of_fasta_names <- list_of_fasta_names[grepl(".FASTA", list_of_fasta_names)]
list_of_fasta_names <- gsub(".FASTA","",list_of_fasta_names)

# read all fasta
fasta_combined <- list()
for(i in 1:length(list_of_fasta)){
  fasta <- read.fasta(list_of_fasta[i], seqtype = "AA", as.string = TRUE, whole.header = TRUE)  
  fasta_combined[[list_of_fasta_names[[i]]]] <- fasta[[1]]
}
fasta_combined <- data.frame(Accession = list_of_fasta_names, Accession_Sequence = unlist(fasta_combined))

# left join 
trinity_hmmer <- trinity_hmmer %>% left_join(fasta_combined)

# change order and column names
trinity_hmmer <- trinity_hmmer[,c("ORF", "DataType", "tlen", "start","end", "Acc","Sample", "Hmm", "HmmClass", "qlen","strand","Eval", "score",
                                  "Cov_seq", "Cov_hmm", "Accession", "CasClass", "CasClassSubtype", "ORF_sequence", "Accession_Sequence")]
colnames(trinity_hmmer) <- c("ORF", "DataType", "ORF_length", "ORF_start","ORF_end", "Read","Sample", "Hmm", "HmmClass", "Hmm_length","strand","Eval", "score",
                             "Coverage_seq", "Coverage_Hmm", "Accession", "CasClass", "CasClassSubtype", "ORF_sequence", "Accession_Sequence")

####
# Filter CCTyper Results ####
####

trinity_hmmer_filtered <- as_tibble(trinity_hmmer) %>% filter(ORF_length > 50, Eval < 0.01)

####
# Cluster ORFs ####
####

# These clusters were found using CD-HIT clustering with fasta input Trinity_sense_hmmer_filtered.fasta 
ORF_Clusters <- read.table("ORF_Clustering/clusters_ORF.txt", sep = "\t", header = TRUE)
trinity_hmmer_filtered <- trinity_hmmer_filtered %>% left_join(ORF_Clusters, by = c("ORF" = "ORF"))

####
# Merge CCTyper results and BLAST Results ####
####

# merge RNA and DNA BLAST result tables
Blast_DNA_results <- readRDS("BLAST/DNA/Trinity_DNA_sense_proteins_filtered_cas_merged.rds")
Blast_DNA_results$QueryID <- gsub("TRINITY", "TRINITY_DNA", Blast_DNA_results$QueryID)
Blast_RNA_results <- readRDS("BLAST/RNA/Trinity_RNA_sense_proteins_filtered_cas_merged.rds")
Blast_RNA_results$QueryID <- gsub("TRINITY", "TRINITY_RNA", Blast_RNA_results$QueryID)
Blast_results <- rbind(Blast_RNA_results, Blast_DNA_results)
Blast_results <- Blast_results[,c("QueryID", "SubjectID", "Alignment.Length", "Perc.Ident", "E")]
colnames(Blast_results) <- paste0("BLAST_", colnames(Blast_results))

# filter NR BLAST alignments and get the best match 
Blast_results_filtered <- as_tibble(Blast_results) %>% 
  filter(BLAST_E < 0.01, BLAST_Alignment.Length > 30) %>% 
  group_by(BLAST_QueryID) %>% 
  slice_max(order_by = BLAST_Perc.Ident, n = 1) %>% 
  group_by(BLAST_QueryID) %>% 
  mutate(BLAST_SubjectID = paste0(BLAST_SubjectID, collapse = "|")) %>% 
  mutate(min_BLAST_AL = min(BLAST_Alignment.Length)) %>% 
  mutate(max_BLAST_E = max(BLAST_E)) %>% 
  select(-c(BLAST_E, BLAST_Alignment.Length)) %>% 
  distinct()

# merge with BLAST
trinity_hmmer_filtered_withBLAST <- trinity_hmmer_filtered %>% left_join(Blast_results_filtered, by = c("ORF" = "BLAST_QueryID"))

####
# Merge CCTyper+BLAST results with RNA Replicase (BLAST) search results ####
####

# get RNA and DNA results
Blast_DNA_results_RNArep <- readRDS("BLAST/DNA/Trinity_DNA_sense_proteins_filtered_cas_BLAST_RNA_replicase.rds")
Blast_DNA_results_RNArep$QueryID <- gsub("^", "TRINITY_DNA_DN", Blast_DNA_results_RNArep$QueryID)
Blast_RNA_results_RNArep <- readRDS("BLAST/RNA/Trinity_RNA_sense_proteins_filtered_cas_BLAST_RNA_replicase.rds")
Blast_RNA_results_RNArep$QueryID <- gsub("TRINITY", "TRINITY_RNA", Blast_RNA_results_RNArep$QueryID)
Blast_results_RNArep <- rbind(Blast_RNA_results_RNArep, Blast_DNA_results_RNArep)
Blast_results_RNArep <- Blast_results_RNArep[,c("QueryID", "Perc.Ident", "Alignment.Length", "E")]
colnames(Blast_results_RNArep) <- paste0("BLAST_RNArep_", colnames(Blast_results_RNArep))

# filter BLAST RNA Replicase alignments
Blast_results_RNArep_filtered <- as_tibble(Blast_results_RNArep) %>% 
  filter(BLAST_RNArep_E < 0.01, BLAST_RNArep_Alignment.Length > 30)

# merge with RNA replicase 
trinity_hmmer_filtered_withBLASTandRNArep <- trinity_hmmer_filtered_withBLAST %>% 
  left_join(Blast_results_RNArep_filtered, by = c("ORF" = "BLAST_RNArep_QueryID")) %>% 
  select(-c(Accession_Sequence))

####
# Sort Results for each Representative ORF ####
####

# get representative ORFs for both RNA and DNA
trinity_hmmer_filtered_withBLASTandRNArep_RepresentORF <- 
  trinity_hmmer_filtered_withBLASTandRNArep %>%
  group_by(Cluster, DataType) %>% summarise(RepresentORF = ORF[which.max(ORF_length)]) %>% 
  group_by(Cluster) %>% mutate(allORFs = paste0(RepresentORF, collapse = "|")) %>% 
  left_join(distinct(trinity_hmmer_filtered_withBLASTandRNArep[,c("ORF", "ORF_length")]), by = c("RepresentORF" = "ORF")) %>% 
  group_by(Cluster) %>% summarise(RepresentORF = RepresentORF[which.max(ORF_length)], 
                                  allORFs = allORFs[1])

# filter for representative ORFs
trinity_hmmer_filtered_withBLASTandRNArep_clustered <- 
  trinity_hmmer_filtered_withBLASTandRNArep[trinity_hmmer_filtered_withBLASTandRNArep$ORF %in% trinity_hmmer_filtered_withBLASTandRNArep_RepresentORF$RepresentORF,]
trinity_hmmer_filtered_withBLASTandRNArep_clustered <- trinity_hmmer_filtered_withBLASTandRNArep_clustered %>%
  left_join(trinity_hmmer_filtered_withBLASTandRNArep_RepresentORF, by = c("ORF" = "RepresentORF"))

# group by representative ORFs
trinity_hmmer_filtered_withBLASTandRNArep_clustered <- 
  trinity_hmmer_filtered_withBLASTandRNArep_clustered %>%
  group_by(ORF) %>% 
  mutate(AllHmm = paste(paste0(Hmm, " (", Hmm_length, ")")[order(Eval)], collapse = " | ")) %>% 
  select(c(ORF, DataType, allORFs, ORF_length, ORF_sequence, AllHmm, BLAST_SubjectID, BLAST_Perc.Ident, min_BLAST_AL, BLAST_RNArep_Perc.Ident, BLAST_RNArep_Alignment.Length)) %>% 
  distinct()

# write table
class(trinity_hmmer_filtered_withBLASTandRNArep_clustered) <- "data.frame"
write.csv(trinity_hmmer_filtered_withBLASTandRNArep_clustered, file = "trinity_hmmer_filtered_withBLASTandRNArep.csv", row.names = FALSE)

# write ORFs to fasta file 
write.fasta(sequences = as.list(trinity_hmmer_filtered_withBLASTandRNArep_clustered$ORF_sequence), 
            names = trinity_hmmer_filtered_withBLASTandRNArep_clustered$ORF, 
            file.out = "trinity_hmmer_filtered_withBLASTandRNArep.fasta")